topics/user_interface_components/status_bar_widgets.md

```diff
@@ -0,0 +1,124 @@
+To implement it, override `ID()` where returns the unique ID of the widget.
+This identifier may be needed to later get a widget instance.
+
+The IntelliJ Platform provides several predefined widget appearance options:
+
+- `com.intellij.openapi.wm.StatusBarWidget.IconPresentation`
+
+  Widget with only an icon.
+
+  Example:
+  [PowerSaveStatusWidgetFactory](upsource:///platform/platform-impl/src/com/intellij/openapi/wm/impl/status/PowerSaveStatusWidgetFactory.java)
+
+- `com.intellij.openapi.wm.StatusBarWidget.TextPresentation`
+
+  Widget with only a text.
+
+  Example:
+  [PositionPanel](upsource:///platform/platform-impl/src/com/intellij/openapi/wm/impl/status/PositionPanel.java)
+
+- `com.intellij.openapi.wm.StatusBarWidget.MultipleTextValuesPresentation`
+
+  Widget with a text and a popup.
+
+  Example:
+  [DvcsStatusWidget](upsource:///platform/dvcs-impl/src/com/intellij/dvcs/ui/DvcsStatusWidget.java)
+
+> Note that they can't be combined to get, for example, an icon and a text.
+>
+{type="note"}
+
+To use the selected appearance, return a class that implements one of the above interfaces from `getPresentation()`.
+
+To create a widget with custom content, it should implement the
+[`CustomStatusBarWidget`](upsource:///platform/platform-api/src/com/intellij/openapi/wm/CustomStatusBarWidget.java)
+interface.
+Override `getComponent()` to return the custom widget's component to display.
+
+Example:
+[MemoryUsagePanel](upsource:///platform/platform-impl/src/com/intellij/openapi/wm/impl/status/MemoryUsagePanel.java)
+
+## `EditorBasedStatusBarPopup`
+
+`EditorBasedStatusBarPopup` is the basis for all widgets that have a popup with a list of actions.
+For example, the encoding widget of the current file.
+
+First, implement `createComponent()`, where returns a component for display.
+Each update of the widget IDE calls `updateComponent()` to update this component.
+In `updateComponent()` implementation, you can describe how the widget should change depending on the current state.
+
+Implement `getWidgetState()` where returns the current state of the widget.
+This state will be passed to the `updateComponent()` when the widget is updated.
+The method accepts a file that's now open in the editor.
+To create your own state class, inherit it from `com.intellij.openapi.wm.impl.status.EditorBasedStatusBarPopup.WidgetState.WidgetState`.
+
+Implement `ID()`, where returns the unique ID of the widget.
+This identifier may be needed to later get a widget instance.
+
+Implement `createInstance()`, where returns the new widget instance.
+
+Finally, implement the `createPopup()` method where returns the [popup](popups.md) that will be displayed when the widget is clicked.
+
+Also, it can be useful to override `registerCustomListeners()` to add custom listeners.
+This can be useful if you want to set custom listeners for a widget that will update it.
+
+To update a widget, use `update()`.
+
+## How to change visibility programmatically
```

Comment:
I suggest changing questions to regular titles as this is not an FAQ page.
Example:
"Controlling Widgets Programmatically" or similar and I would merge the next "How to update widget programmatically" to this new section.